### PR TITLE
refactor custom_vjp use for type agreement

### PR DESCRIPTION
**Pull request by mattjj**
When defining a forward rule for `jax.custom_vjp(primal_fn)`, if `primal_fn` has output type `T` then we need the forward rule to have output type `tuple[T, R]` for some `R`. That is, we need the first output of the forward rule to look like the full output of `primal_fn`. (Here the `R` values represent the 'residuals' computed on the forward pass to save for use on the backward pass.)

This PR fixes a disagreement between `custom_vjp`-decorated functions and their corresponding forward rules. In particular, I applied the `custom_vjp` decorator to the functions which only have the primal outputs of interest (and not the extra residual/intermediate values like `row_sum`). (This choice means that our custom rule won't apply at higher orders of differentiation (since the rule doesn't itself call back into the `custom_vjp`-decorated function); but to make the higher-order case work we'd either need to not share the computation of residuals like `row_sum` with the primal computation (maybe XLA could DCE them?), or else we'd need to write a backward rule which handles nonzero gradients with respect to those intermediate outputs.)

The disagreement caused some interesting behavior! Discussed on https://github.com/lucidrains/flash-attention-jax/issues/7

Separately, I'm going to try to get JAX to raise a better error message in this case; the error message was some really confusing JAX-internals thing.

### CODE REVIEW

**Comment by lucidrains** [comment]
@mattjj thank you Matthew!
